packages/exporter-trace-otlp-http/src/platform/node/util.ts

```diff
@@ -76,9 +77,16 @@ export function sendWithHttp<ExportItem, ServiceRequest>(
     });
   });
 
+  req.setTimeout(exporterTimeout, function requestTimeout () {
```

Comment:
I don't think that this does actually what is expected there for following reasons:
* according to [HTTP docs](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#requestsettimeouttimeout-callback) the time until a socket is assigned is not included here
* once a socket is assigned it measures socket inactivity according to [NET docs](https://nodejs.org/dist/latest-v16.x/docs/api/net.html#socketsettimeouttimeout-callback)

So if a connection is terrible slow but alive no timeout will happen.

This is different to the timeout option in XHR which specifies how long a request may take.

I think we need a dedicated timer here instead of using the `setTimeout` API from HTTP.

Reply:
I can replace with a general timer that uses the exporter timeout:

```js 
setTimeout(() => { 
  // don't wait anymore for export 
  onError(new Error('Timeout')); 
}, exporterTimeout); 
```
    
And then set a timeout on the req with `req.setTimeout` to handle request timeouts and set [`options.timeout`](https://nodejs.org/api/http.html#httprequesturl-options-callback) which from my understanding will set the timeout before a socket is connected.

Reply:
`options.timeout` and [agent.option.timeout](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#new-agentoptions) are actually just other ways to call `request.timeout()` - it's done internally.

But to my understanding this timeout measures socket inactivity, not connection setup time, time for sending body (on slow connections,...),...

I think the correct approach is to just start a timer and abort/destroy the outgoing request. Clearly the time should be stopped once the request finished.

Reply:
I did a quick look into java. They seem to [use](https://github.com/open-telemetry/opentelemetry-java/blob/87f788a3b46fc124910f77a516358319cc0ee75f/exporters/otlp/common/src/main/java/io/opentelemetry/exporter/internal/okhttp/OkHttpExporterBuilder.java#L97-L99) the `OkHttpClient` which offers a [`callTimeout`](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/-builder/call-timeout/) API. The timeout given there is a timeout for the complete request, similar as in XHR.

in Node.js `request.setTimeout` is delegated to `socket.setTimeout` which is a socket inactivity timeout. It may fire as expected in some cases but in others it doesn't.

Here a sample where it never fires because socket is continously doing something:
```js
const http = require("http");

const start = Date.now();
const port = 8000;

function log(...args) {
    console.log(Date.now() - start, ...args);
}

http.createServer((req, res) => {
    log("request start");
    const i = setInterval(() => {
        log("read");
        const data = req.read(100);
        if (!data) {
            clearInterval(i);
            res.end();
        }
    }, 100)
}).listen(port, "localhost");

const req = http.request({ port, timeout: 3000, method: "POST" }, (res) => {
    log("response received");
});
req.write(Buffer.alloc(1000));
setInterval(() => req.write(Buffer.alloc(1000)), 1000);
req.on("timeout", () => log("http Timeout"));
```

If the way how exporter uses http can ensure that such thingy is never happening then fine.

Reply:
Ok. So if we use a general timer and destroy the req after a few seconds then it'll destroy the request even if the socket is still in connection setup phase and we would still have time to retry the request again?

Here's a quick mock up of how this general request timer can work with the exporter timer. Sorry for the long code example. 

```js
// example retry attempts and backoff
const DEFAULT_MAX_ATTEMPTS = 4;
const DEFAULT_INITIAL_BACKOFF = 1000;
const DEFAULT_BACKOFF_MULTIPLIER = 1.5;

function sendWithHttp(collector, data, contentType, onSuccess, onError) {
    let retryTimer;

   // reject promise since exporter had 10 secs to export
   // clear retryTimer to prevent possible queued retry requests
    let exporterTimer = setTimeout(() => {
        clearTimeout(retryTimer);
        onError(new Error('Request Timeout'));
    }, collector._timeoutMillis);

    const parsedUrl = new url.URL(collector.url);
    const options = {
        hostname: parsedUrl.hostname,
        port: parsedUrl.port,
        path: parsedUrl.pathname,
        method: 'POST',
        headers: Object.assign({ 'Content-Length': Buffer.byteLength(data), 'Content-Type': contentType }, collector.headers),
        agent: collector.agent
    };

    const request = parsedUrl.protocol === 'http:' ? http.request : https.request;

    const sendWithRetry = (retries = DEFAULT_MAX_ATTEMPTS, backoffMillis = DEFAULT_INITIAL_BACKOFF) => {
        let reqIsDestroyed;

        // general 2 seconds timer for each request
        // req.destroy() will emit an error
        let oneRequestTimer = setTimeout(() => {
            reqIsDestroyed = true;
            req.destroy();
        }, 2000);

        const req = request(options, (res) => {
            let responseData = '';
            res.on('data', chunk => (responseData += chunk));
            res.on('end', () => {
    
                if (res.statusCode && res.statusCode < 299) {
                    api_1.diag.debug(`statusCode: ${res.statusCode}`, responseData);
                    
                    // clear all timers since request was successful and resolve promise
                    clearTimeout(oneRequestTimer);
                    clearTimeout(exporterTimer);
                    clearTimeout(retryTimer);
                    onSuccess();
                } else if (isRetryable(res.statusCode) && retries > 0) {
                    // clear request 2 second timer and try request again
                    clearTimeout(oneRequestTimer);

                    retryTimer = setTimeout(() => {
                        // try request again with backoff
                        return sendWithRetry(retries - 1, backoffMillis * DEFAULT_BACKOFF_MULTIPLIER);
                    }, backoffMillis)
                } else {
                    const error = new otlpTypes.OTLPExporterError(
                        res.statusMessage,
                        res.statusCode,
                        responseData
                    );
                    
                    // clear all timers since request failed and there are no more retries left
                    // then reject promise
                    clearTimeout(oneRequestTimer);
                    clearTimeout(exporterTimer);
                    clearTimeout(retryTimer);
                    onError(error);
                }
            });
        });
    
        req.on('error', (error) => {
            // req was destroyed due to timeout, try again if there are remaining retries
            if (error.code === 'ECONNRESET' && reqIsDestroyed && retries > 0) {
                
                retryTimer = setTimeout(() => {
                    // try request again with backoff
                    return sendWithRetry(retries - 1, backoffMillis * DEFAULT_BACKOFF_MULTIPLIER);
                }, backoffMillis)
                
             // if something else caused an error reject promise
            } else {
                onError(error);
            }
        });
    
        switch (collector.compression) {
            case types_1.CompressionAlgorithm.GZIP: {
                if (!gzip) {
                    gzip = zlib.createGzip();
                }
                req.setHeader('Content-Encoding', 'gzip');
                const dataStream = readableFromBuffer(data);
                dataStream.on('error', onError)
                    .pipe(gzip).on('error', onError)
                    .pipe(req);
                break;
            }
            default:
                req.write(data);
                req.end();
                break;
        }
    }
    sendWithRetry();
}

function isRetryable(statusCode) {
    const retryCodes = [429, 502, 503, 504];
  
    return retryCodes.includes(statusCode);
}
```

I tested this locally (so far) on a sample app with a socket hang up issue and was able to retry the request 3 times (with the current retry backoff strategy). 

Thoughts on this?

Reply:
I think this could work. In general it's hard to follow node HTTP requests in all special cases and there may be subtle differences between the node versions.
It may be needed to listen also on the `error` event of response object as it seems that errors are emitted there once connection was successful but the data phase takes to long (see [here](https://nodejs.org/dist/latest-v16.x/docs/api/http.html#httprequesturl-options-callback), scroll a bit down for the use cases).

I'm also not sure if `error.code === 'ECONNRESET'` is true in all cases.